File: frontend/src/utils/matchmaker.js
import axios from 'axios';

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000';

/* The AI Matchmaker turns one natural-language prompt into the same 9
   structured answers the lifestyle quiz produces, then the RIGHTBREED
   engine scores breeds identically for both paths.

   Primary: backend /api/matchmaker/parse (Claude).
   Fallback: local keyword parser, so the feature always works. */

const has = (text, words) => words.some((w) => text.includes(w));

export function parsePromptLocally(prompt) {
  const t = (prompt || '').toLowerCase();
  const answers = {};

  // Purpose
  const purposes = [];
  if (has(t, ['guard', 'protect', 'security', 'watch dog', 'watchdog'])) purposes.push('guard');
  if (has(t, ['run', 'jog', 'hik', 'trek', 'adventure', 'workout', 'cycling'])) purposes.push('active');
  if (has(t, ['emotional', 'support', 'therapy', 'comfort', 'lonely', 'anxiety', 'stress'])) purposes.push('therapy');
  if (has(t, ['senior citizen', 'elderly parents', 'aged parents', 'retired'])) purposes.push('seniors');
  answers.purpose = purposes.length ? purposes : ['family'];

  // Home
  if (has(t, ['1bhk', '1 bhk', 'studio', 'small flat', 'tiny apartment', 'small apartment'])) answers.home = ['apt-1bhk'];
  else if (has(t, ['3bhk', '3 bhk', '4bhk', '4 bhk', 'big apartment', 'large flat'])) answers.home = ['apt-3bhk'];
  else if (has(t, ['yard', 'garden', 'lawn', 'farmhouse', 'villa'])) answers.home = ['house-yard'];
  else if (has(t, ['independent house', 'bungalow', 'duplex'])) answers.home = ['house-no-yard'];
  else if (has(t, ['2bhk', '2 bhk', 'apartment', 'flat'])) answers.home = ['apt-2bhk'];

  // Climate — city names first, then adjectives
  if (has(t, ['mumbai', 'chennai', 'kolkata', 'kochi', 'goa', 'hyderabad', 'humid', 'hot city', 'hot weather', 'tropical'])) answers.city = ['hot'];
  else if (has(t, ['shimla', 'dehradun', 'darjeeling', 'srinagar', 'ooty', 'hill station', 'cold', 'chilly'])) answers.city = ['cold'];
  else if (has(t, ['delhi', 'gurgaon', 'gurugram', 'noida', 'jaipur', 'lucknow', 'ahmedabad', 'mixed season'])) answers.city = ['mixed'];
  else if (has(t, ['bangalore', 'bengaluru', 'pune', 'chandigarh', 'mysore', 'pleasant', 'moderate'])) answers.city = ['moderate'];

  // Experience
  if (has(t, ['first time', 'first-time', 'beginner', 'never owned', 'never had', 'new to dogs'])) answers.experience = ['none'];
  else if (has(t, ['experienced', 'owned dogs before', 'had dogs', 'grew up with dogs'])) answers.experience = ['experienced'];

  // Activity
  if (has(t, ['very active', 'running', 'jogging', 'gym', 'trekking', 'hiking', 'high energy lifestyle'])) answers.activity = ['high'];
  else if (has(t, ['lazy', 'couch', 'calm life', 'relaxed', 'busy schedule', 'work long', 'not very active', 'low energy'])) answers.activity = ['relaxed'];

  // Budget
  const rupees = t.match(/(?:rs\.?|₹|inr)\s*([\d,]+)/);
  if (rupees) {
    const amount = parseInt(rupees[1].replace(/,/g, ''), 10);
    if (amount < 5000) answers.budget = ['under5k'];
    else if (amount <= 10000) answers.budget = ['5k-10k'];
    else if (amount <= 20000) answers.budget = ['10k-20k'];
    else answers.budget = ['above20k'];
  } else if (has(t, ['cheap', 'low budget', 'affordable', 'low cost', 'budget friendly'])) {
    answers.budget = ['under5k'];
  } else if (has(t, ['premium', 'luxury', 'money is not', 'no budget limit'])) {
    answers.budget = ['above20k'];
  }

  // Family
  const kids = has(t, ['kid', 'child', 'baby', 'toddler', 'son', 'daughter']);
  const seniors = has(t, ['senior', 'elder', 'grandmother', 'grandfather', 'grandparents', 'old parents', 'aged']);
  if (kids && seniors) answers.family = ['both'];
  else if (kids) answers.family = ['kids'];
  else if (seniors) answers.family = ['seniors'];

  // Shedding
  if (has(t, ['no shed', 'low shed', 'hypoallergenic', 'allerg', 'hate hair', 'clean home', "doesn't shed", 'less hair'])) answers.shedding = ['low'];

  // Coat
  if (has(t, ['fluffy', 'long hair', 'long coat', 'furry'])) answers.hair = ['long'];
  else if (has(t, ['short hair', 'short coat', 'sleek', 'easy grooming'])) answers.hair = ['short'];

  return answers;
}

/** Parse with Claude on the backend; falls back to the local parser.
    Returns { answers, source: 'ai' | 'local' }. */
export async function parsePrompt(prompt) {
  try {
    const res = await axios.post(`${API_URL}/api/matchmaker/parse`, { prompt }, { timeout: 20000 });
    if (res.data?.answers && Object.keys(res.data.answers).length) {
      return { answers: res.data.answers, source: res.data.source || 'ai' };
    }
  } catch (err) {
    console.warn('AI matchmaker unavailable, using local parser', err?.message);
  }
  return { answers: parsePromptLocally(prompt), source: 'local' };
}

File: backend/app/routes/matchmaker.py
"""AI Matchmaker: Claude parses a natural-language prompt into the same 9
structured answers the lifestyle quiz produces. The frontend then runs the
RIGHTBREED scoring engine on those answers, so quiz and AI recommendations
share one code path.

The API key stays server-side (ANTHROPIC_API_KEY in backend/.env). When no
key is configured, the route answers with an empty payload and the frontend
falls back to its local keyword parser.
"""

from fastapi import APIRouter
from pydantic import BaseModel

from app.config.settings import settings

try:
    from anthropic import AsyncAnthropic
except ImportError:  # anthropic not installed yet — fallback mode
    AsyncAnthropic = None

router = APIRouter()

MODEL = "claude-haiku-4-5"

SYSTEM_PROMPT = """You extract structured dog-matchmaking preferences from a prospective Indian dog owner's message.

Map what the user says onto the answer values. Only commit to a value the message actually supports; use "unknown" (or an empty array) when the message says nothing about that dimension. Notes:
- city refers to climate. Indian city hints: Mumbai/Chennai/Kolkata/Kochi/Goa/Hyderabad = hot; Bangalore/Pune/Chandigarh/Mysore = moderate; Delhi/Jaipur/Lucknow/Ahmedabad = mixed; hill stations (Shimla, Ooty, Darjeeling, Srinagar) = cold.
- budget is monthly care budget in INR.
- family describes who lives at home (kids / seniors / both / adults only).
- shedding: "low" means the user wants minimal shedding; "fine" means they explicitly don't mind hair.
- purpose may have several values; "therapy" covers emotional support and companionship for loneliness/stress."""

ANSWERS_SCHEMA = {
    "type": "object",
    "properties": {
        "purpose": {
            "type": "array",
            "items": {"type": "string", "enum": ["family", "guard", "active", "therapy", "seniors"]},
        },
        "home": {
            "type": "string",
            "enum": ["apt-1bhk", "apt-2bhk", "apt-3bhk", "house-no-yard", "house-yard", "unknown"],
        },
        "city": {"type": "string", "enum": ["hot", "moderate", "mixed", "cold", "unknown"]},
        "experience": {"type": "string", "enum": ["none", "some", "experienced", "unknown"]},
        "activity": {
            "type": "array",
            "items": {"type": "string", "enum": ["relaxed", "moderate", "high"]},
        },
        "budget": {"type": "string", "enum": ["under5k", "5k-10k", "10k-20k", "above20k", "unknown"]},
        "family": {"type": "string", "enum": ["kids", "seniors", "both", "adults", "unknown"]},
        "shedding": {"type": "string", "enum": ["fine", "moderate", "low", "unknown"]},
        "hair": {
            "type": "array",
            "items": {"type": "string", "enum": ["short", "medium", "long", "any"]},
        },
    },
    "required": ["purpose", "home", "city", "experience", "activity", "budget", "family", "shedding", "hair"],
    "additionalProperties": False,
}


class MatchmakerRequest(BaseModel):
    prompt: str


@router.post("/parse")
async def parse_prompt(request: MatchmakerRequest):
    api_key = settings.ANTHROPIC_API_KEY
    if not api_key or AsyncAnthropic is None:
        # No key configured — frontend uses its local parser
        return {"answers": {}, "source": "none"}

    try:
        client = AsyncAnthropic(api_key=api_key)
        response = await client.messages.create(
            model=MODEL,
            max_tokens=1024,
            system=SYSTEM_PROMPT,
            messages=[{"role": "user", "content": request.prompt[:2000]}],
            output_config={"format": {"type": "json_schema", "schema": ANSWERS_SCHEMA}},
        )

        import json

        text = next((b.text for b in response.content if b.type == "text"), "{}")
        raw = json.loads(text)

        # Drop unknowns and normalise everything to arrays (quiz answer shape)
        answers = {}
        for key, value in raw.items():
            if isinstance(value, list):
                if value:
                    answers[key] = value
            elif value and value != "unknown":
                answers[key] = [value]

        return {"answers": answers, "source": "ai"}
    except Exception as exc:  # any API failure → frontend local fallback
        print(f"Matchmaker parse failed: {exc}")
        return {"answers": {}, "source": "error"}
